File: StudyNotes/DataStructure/Other/Queue_.swift
//
//  Queue.swift
//  DataStructure
//
//  Created by 朱双泉 on 04/01/2018.
//  Copyright © 2018 Castie!. All rights reserved.
//

import Foundation

class Queue_<Element> {
    
    private var capacity: Int
    private lazy var head: Int = 0
    private lazy var tail: Int = 0
    private lazy var length: Int = 0
    private lazy var queue: [Element] = [Element]()
    
    init(_ queueCapacity: Int = 16) {
        capacity = queueCapacity
    }
    
    func clear() {
        head = 0
        tail = 0
        length = 0
    }
    
    func isEmpty() -> Bool {
        return length == 0 ? true : false
    }
    
    func isFull() -> Bool {
        return length == capacity ? true : false
    }
    
    func size() -> Int {
        return length
    }
    
    @discardableResult func entry(_ element: Element) -> Bool {
        guard !isFull() else { return false }
        queue.append(element)
        queue[tail] = element
        tail += 1
        tail %= capacity
        length += 1
        return true
    }
    
    @discardableResult func depart() -> Element? {
        guard !isEmpty() else { return nil }
        let element = queue[head]
        head += 1
        head %= capacity
        length -= 1
        return element
    }
    
    func traverse() {
        print("︵")
        for i in head..<length + head {
            print(queue[i % capacity])
        }
        print("︶")
    }
}

File: StudyNotes/GeekBang.org/专栏/数据结构与算法之美/54 | 算法实战（三）：剖析高性能队列Disruptor背后的数据结构和算法/src/queue/Queue.java
package queue;

public class Queue {
    private Long[] data;
    private int size = 0, head = 0, tail = 0;
    public Queue(int size) {
        this.data = new Long[size];
        this.size = size;
    }

    public boolean add(Long element) {
        if ((tail + 1) % size == head) return false;
        data[tail] = element;
        tail = (tail + 1) % size;
        return true;
    }

    public Long poll() {
        if (head == tail) return null;
        long ret = data[head];
        head = (head + 1) % size;
        return ret;
    }
}
